File: app/src/main/java/com/example/android/sunshine/ForecastAdapter.java
package com.example.android.sunshine;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

public class ForecastAdapter extends RecyclerView.Adapter<ForecastAdapter.ForecastAdapterViewHolder> {

    private String[] mWeatherData;
    private final ForecastAdapterOnClickHandler mHandler;

    public ForecastAdapter(ForecastAdapterOnClickHandler mHandler) {
        super();
        this.mHandler = mHandler;
    }

    @NonNull
    @Override
    public ForecastAdapter.ForecastAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        Context context = parent.getContext();
        int  listItemId = R.layout.forecast_list_item;
        LayoutInflater inflater = LayoutInflater.from(context);
        View view = inflater.inflate(listItemId, parent, false);

        return new ForecastAdapterViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull ForecastAdapterViewHolder holder, int position) {
        holder.mWeatherTextView.setText(this.mWeatherData[position]);
    }

    @Override
    public int getItemCount() {
        if(mWeatherData == null)
            return 0;
        else
            return mWeatherData.length;
    }

    public void setWeatherData(String[] weatherData) {
        this.mWeatherData = weatherData;
        notifyDataSetChanged();
    }

    /* Implement OnClickListener to respond to click event*/
    public class ForecastAdapterViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
        public final TextView mWeatherTextView;

        public ForecastAdapterViewHolder(@NonNull View itemView) {
            super(itemView);
            mWeatherTextView = itemView.findViewById(R.id.tv_weather_data);
            itemView.setOnClickListener(this);
        }

        @Override
        public void onClick(View view) {
            int position = getAdapterPosition();
            String weather = mWeatherData[position];
            mHandler.onClick(weather);
        }
    }

    public interface ForecastAdapterOnClickHandler{
        public void onClick(String weather);
    }
}

File: app/src/main/java/com/example/android/sunshine/ForecastAdapter.kt
package com.example.android.sunshine

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.lifecycle.LifecycleOwner
import androidx.lifecycle.LiveData
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.example.android.sunshine.ForecastAdapter.ForecastAdapterViewHolder

class ForecastAdapter(private val mHandler: ForecastAdapterOnClickHandler) : RecyclerView.Adapter<ForecastAdapterViewHolder>() {

    private var mWeatherData = arrayOf<String>()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ForecastAdapterViewHolder {
        val context = parent.context
        val listItemId = R.layout.forecast_list_item
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(listItemId, parent, false)
        return ForecastAdapterViewHolder(view)
    }

    override fun onBindViewHolder(holder: ForecastAdapterViewHolder, position: Int) {
        holder.mWeatherTextView.text = mWeatherData[position]
    }

    override fun getItemCount(): Int {
        return mWeatherData.size
    }

    fun setWeatherData(weatherData: Array<String>) {
        mWeatherData = weatherData
        notifyDataSetChanged()
    }

    /* Implement OnClickListener to respond to click event*/
    inner class ForecastAdapterViewHolder(itemView: View) : ViewHolder(itemView), View.OnClickListener {
        val mWeatherTextView: TextView = itemView.findViewById(R.id.tv_weather_data)
        override fun onClick(view: View) {
            val position = adapterPosition
            val weather = mWeatherData[position]
            mHandler.onClick(weather)
        }

        init {
            itemView.setOnClickListener(this)
        }
    }

    interface ForecastAdapterOnClickHandler {
        fun onClick(weather: String?)
    }
}